[add] rutracker.org topic pages urlrewrite

Niluge-KiWi opened the pull request:
### Motivation for changes:
Get magnet url from rutracker.org topics.

### Detailed changes:
- get torrent hash from topic id: uses https://api.t-ru.org to get topic data
- use it to reconstruct a magnet link

### Config usage if relevant (new plugin or updated schema):
rutracker.org has an atom feed: http://feed.rutracker.cc/atom/f/0.atom 

### Log and/or tests output (preferably both):
I added a test in `tests/test_urlrewriting.py`. Not sure it's the best way to do it (and it requires online access: don't know if we want to mock the network call nor how to do it.

#### To Do:

- [ ] maybe deprecate the old `rutracker_auth` plugin: I'm not sure it's that useful since torrents are public there, and my plugin creates magnet link without needing to really download `.torrent` files from rutracker.org.
- [ ] maybe rename the file from `site_rutracker.py` to `rutracker.py` if the old plugin is removed, or merge both?

githubbla commented:
Could you please provide sample how to configure this staff?

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,55 @@
+    # urlrewriter API
+    def url_rewritable(self, task, entry):
+        url = entry['url']
+        if url.startswith('https://rutracker.org/forum/viewtopic.php?t='):
+            return True
+        return False
```
more pythonic to do this:
```suggestion
        return url.startswith('https://rutracker.org/forum/viewtopic.php?t=')
```

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,55 @@
+        """
+
+        url = entry['url']
+        logger.info('rewriting download url: %s' % url)
```
please use curly braces for loggers:
```suggestion
        logger.info('rewriting download url: {}', url)
```

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,55 @@
+
+        topic_id = parse_qs(urlparse(url).query)['t'][0]
+
+        api_url = self.base_url + '/v1/get_tor_topic_data?by=topic_id&val=%s' % topic_id
```
please use fstrings

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,55 @@
+
+        topic = topic_request.json()['result'][topic_id]
+
+        magnet_url = 'magnet:?xt=urn:btih:' + topic['info_hash'] + '&tr=http%3A%2F%2Fbt.t-ru.org%2Fann%3Fmagnet'
```
please use fstrings

Niluge-KiWi commented:
Thanks @liiight for the review, I updated my PR with all your suggestions, and also added more features: 
- multi tracker (they seem to randomly generate one in 4, I generate all 4 (easier for the test, maybe I should random the order and/or follow what upstream does: send only one, but it seems less robust, but then tests are harder to do..))
- topic title as Display Name in magnet uri, hoping it would help with some clients when magnetization timeouts.

If you have time, could you please reply to my questions in my initial comment? Thanks !

Niluge-KiWi commented:
@githubbla It should work using their forum rss/atom feed: http://feed.rutracker.cc/atom/f/0.atom

If you also want music parsing I have a somewhat working version 0 there: https://github.com/Niluge-KiWi/Flexget/tree/dev-music
(usage: filter previous atom feed with an artist list)

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,59 @@
+
+        topic_id = parse_qs(urlparse(url).query)['t'][0]
+
+        api_url = f"{self.base_url}/v1/get_tor_topic_data?by=topic_id&val={topic_id}"
```
Please pass query params as it's own dict

Niluge-KiWi commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,59 @@
+            topic_request = task.requests.get(api_url)
+        except RequestException as e:
+            logger.error('rutracker request failed: {}', e)
+            raise UrlRewritingError('rutracker request failed: %s' % e)
```
I was just following other sites/ coding style, notably `site_1337x.py`. But ok let's modernize at least the new code.

Niluge-KiWi commented:
@liiight If you have time, could you please reply to my questions in my initial comment? Thanks !

liiight commented:
@Niluge-KiWi I'll try to respond ASAP

Niluge-KiWi commented:
No urgency at all!

Le jeu. 11 juin 2020 à 23:30, Or Carmi <notifications@github.com> a écrit :

> @Niluge-KiWi <https://github.com/Niluge-KiWi> I'll try to respond ASAP
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Flexget/Flexget/pull/2609#issuecomment-642937620>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACAIY723RUGLXRRD6GOZUTRWFEJHANCNFSM4KZ6G34Q>
> .
>

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,63 @@
+            logger.debug('requesting: {} params={}', api_url, api_params)
+            topic_request = task.requests.get(api_url, params=api_params)
+        except RequestException as e:
+            logger.error('rutracker request failed: {}', e)
```
Redundant, will be shown by exception

liiight commented on flexget/components/sites/sites/site_rutracker.py:
```diff
@@ -0,0 +1,63 @@
+
+        magnet = {
+            'xt': f"urn:btih:{topic['info_hash']}",
+            'tr': [ 'http://bt{}.t-ru.org/ann?magnet'.format(i) for i in ['', '2', '3', '4'] ],
```
Redundant space char. Also please use fstrings

Niluge-KiWi commented:
@liiight last round of requested changes done!

Side note: most of your (valid) remarks also apply to other files in `sites/*.py` (notably `site_1337.py`), it's thus confusing from an external contributor: I mimic what I see in the project, notably I tried to validate with flake8 initially (as documented there https://flexget.readthedocs.io/en/latest/develop/index.html#code-style) but got many errors running it on `sites/*.py`, so I concluded it's not enforced in the CI/mandatory. Maybe cleaning these and enforcing more things in flake8/pep8/other checker if needed could remove this manual back and forth for 4 months. I do understand it would take time. But once it's enforced by CI checks it won't diverge again: it's a one-time cost. I could help if needed (autopep8/yapf hopefully can lift most of the work).
It would give you more time to answer deeper questions about code architecture than coding style.

liiight commented:
@Niluge-KiWi thank you for your contribution! 
You are absolutely right about the formatting issue. We do use pre-commit which runs black manually, I guess we should maybe just add it as a CI step and let it change and commit on its own. maybe...

n00buk commented:
Unfortunately this completely breaks external RSS feeds with direct torrent links to RuTracker. Can the old rutracker_auth plugin be restored and work in parallel with this?

BrutuZ commented:
@n00buk In that case wouldn't it be fine to just disable the url rewriting for the task?
```yaml
disable:
  - urlrewriting
```

n00buk commented:
@BrutuZ, I might be doing it wrong, but it didn't work. Here is what i tried...

    task-name:
      rss: {? key.rssurl ?}
      accept_all: yes
      rutracker_auth:
        username: \''{? key.rut_u ?}'\'
        password: \''{? key.rut_p ?}'\'
      set:
        path: '{? folder.ong_dl ?}'
      disable: urlrewriting
      transmission:
        host: localhost
        port: 9091
        ratio: -1
        username: '{? key.tr_u ?}'
        password: '{? key.tr_p ?}'

... and what I got ...

    2020-07-08 22:05:47 INFO     transmission  task-name      Trying to connect to transmission...
    2020-07-08 22:05:47 INFO     transmission  task-name      Successfully connected to transmission.
    2020-07-08 22:05:47 VERBOSE  backlog       task-name      Added 3 entries from backlog
    2020-07-08 22:05:47 VERBOSE  details       task-name      Produced 4 entries.
    2020-07-08 22:05:47 VERBOSE  task          task-name      ACCEPTED: `torrent name 1` by accept_all plugin
    2020-07-08 22:05:47 VERBOSE  task          task-name      ACCEPTED: `torrent name 2` by accept_all plugin
    2020-07-08 22:05:47 VERBOSE  task          task-name      ACCEPTED: `torrent name 3` by accept_all plugin
    2020-07-08 22:05:47 VERBOSE  task          task-name      ACCEPTED: `torrent name 4` by accept_all plugin
    2020-07-08 22:05:47 CRITICAL plugin        task-name      Host unreachable.
    2020-07-08 22:05:47 WARNING  task          task-name      Aborting task (plugin: rutracker_auth)
